src/compiler/extensions.ts

```diff
@@ -0,0 +1,379 @@
+                        }
+                    }
+                    catch (e) {
+                        error = e;
+                    }
+                }
+                else if (!host.loadExtension) {
+                    error = new Error("Extension loading not implemented in host!");
+                }
+                if (error) {
+                    diagnostics.push(createCompilerDiagnostic(Diagnostics.Extension_loading_failed_with_error_0, error));
+                }
+                return { name, result, error };
+            });
+            const successfulExtensionLoadResults = filter(extensionLoadResults, res => !res.error);
+            const preparedExtensionObjects = map(successfulExtensionLoadResults, res => {
+                if (res.result) {
+                    return reduceProperties(res.result, (aggregate: Extension[], potentialExtension: BaseProviderStatic, key: string) => {
+                        if (!potentialExtension) {
+                            return; // Avoid errors on explicitly exported null/undefined (why would someone do that, though?)
+                        }
+                        const annotatedKind = potentialExtension["extension-kind"];
+                        if (typeof annotatedKind === "string") {
+                            const ext: ExtensionBase = {
+                                name: key !== "default" ? `${res.name}[${key}]` : res.name,
+                                args: extensionNames === extOptions ? undefined : (extOptions as Map<any>)[res.name],
+                                kind: annotatedKind,
+                            };
+                            switch (ext.kind) {
+                                case ExtensionKind.SemanticLint:
+                                case ExtensionKind.SyntacticLint:
+                                case ExtensionKind.LanguageService:
+                                    if (typeof potentialExtension !== "function") {
+                                        diagnostics.push(createCompilerDiagnostic(
+                                            Diagnostics.Extension_0_exported_member_1_has_extension_kind_2_but_was_type_3_when_type_4_was_expected,
+                                            res.name,
+                                            key,
+                                            (ts as any).ExtensionKind[annotatedKind],
+                                            typeof potentialExtension,
+                                            "function"
+                                        ));
+                                        return;
+                                    }
+                                    (ext as (SemanticLintExtension | SyntacticLintExtension | LanguageServiceExtension)).ctor = potentialExtension as (SemanticLintProviderStatic | SyntacticLintProviderStatic | LanguageServiceProviderStatic);

```

Comment:
Just part of the API. It just happens to be that the top level shape of all three of these extension kinds is identical (Classish) - though nothing requires that!